storage-node/src/cache/data_store_cache/memdisk/mod.rs

```diff
@@ -428,4 +565,103 @@ mod tests {
         }
         assert_eq!(written_data_size, 113629);
     }
+
+    #[tokio::test]
+    async fn test_same_requests_simultaneously_disk() {
+        let tmp = tempfile::tempdir().unwrap();
+        let disk_base_path = tmp.path().to_owned();
+        let cache = MemDiskStoreCache::new(
+            LruReplacer::new(1024 * 512),
+            disk_base_path.to_str().unwrap().to_string(),
+            None,
+            None,
+        );
+        let bucket = "tests-parquet".to_string();
+        let keys = vec!["userdata2.parquet".to_string()];
+        let remote_location = bucket.clone() + &keys[0];
+        let reader = MockS3Reader::new(bucket.clone(), keys.clone()).await;
+        let reader2 = MockS3Reader::new(bucket.clone(), keys.clone()).await;
+        let reader3 = MockS3Reader::new(bucket.clone(), keys).await;
+
+        let put_data_fut_1 = cache.put_data_to_cache(
+            remote_location.clone(),
+            None,
+            reader.into_stream().await.unwrap(),
+        );
+        let put_data_fut_2 = cache.put_data_to_cache(
+            remote_location.clone(),
+            None,
+            reader2.into_stream().await.unwrap(),
+        );
+        let put_data_fut_3 = cache.put_data_to_cache(
+            remote_location.clone(),
+            None,
+            reader3.into_stream().await.unwrap(),
+        );
+
+        let res = join!(put_data_fut_1, put_data_fut_2, put_data_fut_3);
+        assert!(res.0.is_ok());
+        assert!(res.1.is_ok());
+        assert!(res.2.is_ok());
+        assert_eq!(res.0.unwrap(), 112193);
+        assert_eq!(res.1.unwrap(), 112193);
+        assert_eq!(res.2.unwrap(), 112193);
```

Comment:
Per [tokio's document](https://docs.rs/tokio/latest/tokio/macro.join.html#:~:text=By%20running%20all,to%20join!.), the way the test is written here won't run in parallel. In our case, there would be multiple calls to the storage manager at the same time, and the mem-disk cache would handle multiple put/get requests at the same time, which should be in parallel. So we might spawn a number of tokio tasks here and call `join!` on those tasks rather than directly joining on the futures.

Reply:
Good catch!